doc/user/content/sql/patterns/window-functions.md

```diff
@@ -1,8 +1,126 @@
 ---
-title: "Window functions"
-description: "Use lateral joins as a streaming-efficient stand-in for many window functions."
-draft: true
+title: "Window functions (OVER clause)"
+description: "Avoid window functions, e.g., use lateral joins as a streaming-efficient stand-in for many window functions."
 menu:
   main:
     parent: 'sql-patterns'
 ---
+
+Materialize currently does not handle window functions (`OVER` clause) efficiently for large partition sizes: when an input record is added/removed/changed, the system recomputes the results for the entire partition that the changed record belongs to. We are planning to make window functions more efficient in Materialize, but for the time being, this page shows some workarounds for avoiding the use of window functions.
+
+Note that this page **does _not_ refer to temporal windows**, i.e., when a window belongs to a time interval. Examples of temporal windows are when we select only the last hour of an infinite stream of records, or when we group input records by which hour they occurred in. Such queries are efficiently supported in Materialize by using [temporal filters and temporal groupings](/sql/patterns/temporal-filters).
+
+By "window functions", this page means functions such as `LAG`, `LEAD`, `ROW_NUMBER`, `FIRST_VALUE`. You can access these functions with the `OVER` clause, in which you use the `PARTITION BY` clause to group input records into partitions by a key, and use the `ORDER BY` clause to order records inside each partition. Window functions compute results based on the position of records inside these partitions. For example, `x - LAG(x)` computes the differences of attribute `x` between each pair of _consecutive_ rows in each partition.
+
+## Example data
+
+We use the following table as input for later examples queries on this page:
+
+```sql
+CREATE TABLE cities (
+    name text NOT NULL,
+    state text NOT NULL,
+    pop int NOT NULL
+);
+
+INSERT INTO cities VALUES
+    ('Los_Angeles', 'CA', 3979576),
+    ('Phoenix', 'AZ', 1680992),
+    ('Houston', 'TX', 2320268),
+    ('San_Diego', 'CA', 1423851),
+    ('San_Francisco', 'CA', 881549),
+    ('New_York', 'NY', 8336817),
+    ('Dallas', 'TX', 1343573),
+    ('San_Antonio', 'TX', 1547253),
+    ('San_Jose', 'CA', 1021795),
+    ('Chicago', 'IL', 2695598),
+    ('Austin', 'TX', 978908);
+```
+
+## Top K using `ROW_NUMBER`
+
+In other databases, a popular way of computing the top _K_ records per key is to use the `ROW_NUMBER` window function. For example, to get the 3 most populous city in each state:
+```
+SELECT state, name
+FROM (
+  SELECT state, name, ROW_NUMBER() OVER
+    (PARTITION BY state ORDER BY pop) as row_num
+  FROM cities
+)
+WHERE row_num <= 3;
+```
+
+If there are states that have many cities, then it is better for performance to express this query using a lateral join (or `DISTINCT ON` is _K_ = 1) instead of window functions, which we explain in [Top K by group](/sql/patterns/top-k).
```

Comment:
```suggestion
If there are states that have many cities, a more performant way to express this in Materialize is to use a lateral join (or `DISTINCT ON`, if _K_ = 1) instead of window functions. See [Top K by group](/sql/patterns/top-k) for the alternative implementation.
```

Reply:
We should provide the alternative here, rather than point at the Top K page. Or, maybe we should merge the two pages? Just thinking out loud, no need to block this PR on that.